### ISSUE TITLE
Cannot compile model after upgrade to 0.3.1

### ISSUE THREAD

**Issue by magic282** [fchollet/keras#1491]
Hi,

I am using Keras LSTM doing a sequence prediction work. I am trying to use stateful layers so I upgrade Keras to 0.3.1(the latest) from 0.2. But the following code cannot be compiled any more.

Did I miss anything here? Thanks.

    from keras.models import Sequential
    from keras.layers.core import Dense, Activation, Dropout, Masking, TimeDistributedDense
    from keras.layers.recurrent import LSTM
    from keras.regularizers import l2
    
    model = Sequential()
    model.add(Masking(0.0, input_shape=(50, 2000)))
    model.add(LSTM(output_dim=512, return_sequences=True))
    model.add(TimeDistributedDense(output_dim=87, activation='softmax'))
    model.compile(optimizer='adadelta', loss='categorical_crossentropy')

**Comment by magic282** [comment]
update: the code I downloaded on Jan 7 seems to be working.

**Comment by magic282** [comment]
update2:

The Keras 0.3.1(Jan 7th), failed when running:

train_X shape: (33172L, 140L, 2135L)
train_Y shape: (33172L, 140L, 87L)
A LSTM Classifier
Building model...       2016-01-19 18:22:12.823000
Model compiled. 2016-01-19 18:23:01.914000
######################################
Iteration # 0   2016-01-19 18:23:01.915000
Epoch 1/1
Traceback (most recent call last):
    self.lstm_classfier.train(train_X=train_X, train_Y=train_Y, batch_size=Config.lstm_mini_batch_size, epoch=1)
  File "LSTMClassifier.py", line 52, in train
    self.model.fit(X=train_X, y=train_Y, batch_size=batch_size, nb_epoch=epoch, verbose=1, show_accuracy=True)
  File "`\Anaconda2\lib\site-packages\keras\models.py", line 581, in fit
    shuffle=shuffle, metrics=metrics)
  File "`Anaconda2\lib\site-packages\keras\models.py", line 239, in _fit
    outs = f(ins_batch)
  File "`Anaconda2\lib\site-packages\keras\backend\theano_backend.py", line 365, in __call__
    return self.function(*inputs)
  File "`Anaconda2\lib\site-packages\theano\compile\function_module.py", line 871, in __call__
    storage_map=getattr(self.fn, 'storage_map', None))
  File "\Anaconda2\lib\site-packages\theano\gof\link.py", line 314, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "\Anaconda2\lib\site-packages\theano\compile\function_module.py", line 859, in __call__
    outputs = self.fn()
ValueError: GpuElemwise. Input dimension mis-match. Input 1 (indices start at 0) has shape[0] == 140, but the output's size on that axis is 64.
Apply node that caused the error: GpuElemwise{Composite{((i0 * i1 * i2 * i3) / (i4 * i5 * i6 * i6))}}[(0, 0)](GpuFromHost.0, GpuDimShuffle{0,x,x}.0, GpuFromHost.0, GpuDimShuffle{1,
0,2}.0, GpuDimShuffle{x,x,x}.0, GpuElemwise{Clip}[(0, 0)].0, GpuDimShuffle{0,1,x}.0)
Toposort index: 277
Inputs types: [CudaNdarrayType(float32, 3D), CudaNdarrayType(float32, (False, True, True)), CudaNdarrayType(float32, 3D), CudaNdarrayType(float32, 3D), CudaNdarrayType(float32, (Tr
ue, True, True)), CudaNdarrayType(float32, 3D), CudaNdarrayType(float32, (False, False, True))]
Inputs shapes: [(64, 140, 87), (140, 1, 1), (64, 140, 87), (64, 140, 87), (1, 1, 1), (64, 140, 87), (64, 140, 1)]
Inputs strides: [(12180, 87, 1), (1, 0, 0), (12180, 87, 1), (87, 5568, 1), (0, 0, 0), (12180, 87, 1), (140, 1, 0)]
Inputs values: ['not shown', 'not shown', 'not shown', 'not shown', <CudaNdarray object at 0x000000007636F370>, 'not shown', 'not shown']
Outputs clients: [[GpuCAReduce{add}{0,0,1}(GpuElemwise{Composite{((i0 * i1 * i2 * i3) / (i4 * i5 * i6 * i6))}}[(0, 0)].0)]]

HINT: Re-running with most Theano optimization disabled could give you a back-trace of when this node was created. This can be done with by setting the Theano flag 'optimizer=fast_
compile'. If that does not work, Theano optimizations can be disabled with 'optimizer=None'.
HINT: Use the Theano flag 'exception_verbosity=high' for a debugprint and storage map footprint of this apply node.

**Comment by jfsantos** [comment]
I am having similar problems when trying to use an `Embedding` layer with masking.

**Comment by magic282** [comment]
@jfsantos I still cannot compile the model with the following error msg:

Using Theano backend.
Traceback (most recent call last):
  File "cannotcompile.py", line 10, in <module>
    model.compile(optimizer='adadelta', loss='categorical_crossentropy')
  File "D:\Projects\opensource\keras\keras\models.py", line 435, in compile
    self.y_train = self.get_output(train=True)
  File "D:\Projects\opensource\keras\keras\layers\containers.py", line 128, in get_output
    return self.layers[-1].get_output(train)
  File "D:\Projects\opensource\keras\keras\layers\core.py", line 1114, in get_output
    X = self.get_input(train)
  File "D:\Projects\opensource\keras\keras\layers\core.py", line 173, in get_input
    previous_output = self.previous.get_output(train=train)
  File "D:\Projects\opensource\keras\keras\layers\recurrent.py", line 150, in get_output
    mask=mask)
  File "D:\Projects\opensource\keras\keras\backend\theano_backend.py", line 450, in rnn
    mask = mask.dimshuffle(axes)
  File "C:\Anaconda2\lib\site-packages\theano\tensor\var.py", line 360, in dimshuffle
    pattern)
  File "C:\Anaconda2\lib\site-packages\theano\tensor\elemwise.py", line 146, in __init__
    (i, j, len(input_broadcastable)))
ValueError: new_order[2] is 2, but the input only has 2 axes.

**Comment by magic282** [comment]
update3:
I debugged into keras code, the error occurs at line 450 in theano_backend.py:

```Python

    ndim = inputs.ndim
    assert ndim >= 3, "Input should be at least 3D."
    axes = [1, 0] + list(range(2, ndim))
    inputs = inputs.dimshuffle(axes)
    if mask is None:
        mask = expand_dims(ones_like(T.sum(inputs, axis=-1)))
    else:
        mask = mask.dimshuffle(axes)
```

So I guess it may be caused by the masking layer.

**Comment by magic282** [comment]
It seems that this issue got fixed after upgrade to 0.3.2.
